src/utils/peer.ts

```diff
@@ -29,63 +24,92 @@ function sendPlayerName(c: DataConnection) {
+  addListenersToPeer();
+});
+
+function addListenersToPeer() {
   const p = peer.value;
   if (!p) return;
-  p.off("error")
-    .on("error", (e) => handleErrors(e))
-    .off("connection")
-    .on("connection", (c) => handleConnectionToTheHost(c))
-    .off("open")
-    .on("open", () => handleNonHostPeerOpened(p));
-});
+  p.on("error", (e) => handlePeerError(e))
+    .on("connection", (c) => handleReceivedPeerConnection(c))
+    .on("disconnected", () => handlePeerDisconnectedFromTheServer(p))
+    .on("open", () => handlePeerOpened(p));
+}
 
-function handleConnectionToTheHost(c: DataConnection) {
+function handleReceivedPeerConnection(c: DataConnection) {
   if (!isHost.value) return;
-  applyMessageHandler(c);
-  connectionMap.value = new Map([...connectionMap.value, [c.peer, c]]);
-  c.off("open").on("open", () => {
-    console.info(`New player ${c.peer} joined.`);
-    sendPlayerName(c);
-    boardcastPlayerList();
-  });
+  handleReceivedPeerConnectionToTheHost(c);
 }
 
-function handleNonHostPeerOpened(p: Peer) {
+function handlePeerDisconnectedFromTheServer(p: Peer) {
+  // TODO set a counter and limit times to reconnect, then destory it
+  if (!p.open || p.destroyed) return;
+  p.reconnect();
+}
+
+function handlePeerOpened(p: Peer) {
   if (isHost.value) return;
-  const connection = p.connect(hostId.value);
-  connection.on("open", () => {
-    connectionMap.value = new Map([
-      ...connectionMap.value,
-      [connection.peer, connection],
-    ]);
-    sendPlayerName(connection);
-  });
-  applyMessageHandler(connection);
+  handleNonHostPeerOpen(p);
 }
 
-function handleErrors(e: PeerError<string>) {
+function handlePeerError(e: PeerError<string>) {
   switch (e.type) {
     case "unavailable-id":
-      exitRoom();
-      alert(
-        "Room with this name has already been created. You may want to join instead."
-      );
       break;
     default:
       console.error(e);
       break;
   }
 }
 
-// TODO handle disconnections
-// TODO should not allow same player names
+function handleConnectionOpened(c: DataConnection) {
+  if (isHost.value) {
+    console.info(`New player ${c.peer} joined.`);
+    return;
+  }
+  HandleNonHostConnectionOpened(c);
+}
+
+function handleConnectionClosed(c: DataConnection) {
+  if (!isHost.value) {
+    exitRoom();
+    return;
+  }
+  HandleHostConnectionClosed(c);
+}
+
+function handleReceivedPeerConnectionToTheHost(c: DataConnection) {
+  applyMessageHandler(c);
+  connectionMap.value = new Map([...connectionMap.value, [c.peer, c]]);
+  c.on("open", () => {
+    handleConnectionOpened(c);
+  }).on("close", () => handleConnectionClosed(c));
+}
+
+function handleNonHostPeerOpen(p: Peer) {
+  const connection = p.connect(hostId.value);
+  connection
+    .on("open", () => {
+      handleConnectionOpened(connection);
+    })
+    .on("close", () => handleConnectionClosed(connection));
+  applyMessageHandler(connection);
+}
+
+function HandleNonHostConnectionOpened(c: DataConnection) {
```

Comment:
What does this do actually?

Reply:
so basically, handleClientConnectionSucceed

Reply:
Kind of, but `open` is the raw event name and I just used it.